src/runtime_src/core/edge/sw_em/generic_pcie_hal2/shim.cxx

```diff
@@ -840,6 +858,89 @@ namespace xclcpuemhal2 {
     }
     return 0;
   }
+  
+   bool CpuemShim::isValidCu(uint32_t cu_index) {  
+    // get sorted cu addresses to match up with cu_index
+    const auto& cuidx2addr = mCoreDevice->get_cus();
+    if (cu_index >= cuidx2addr.size()) {
+      std::string strMsg = "ERROR: [SW-EMU 20] invalid CU index: " + std::to_string(cu_index);
+      mLogStream << __func__ << strMsg << std::endl;
+      return false;
+    }
+    return true;
+  }
+
+  uint64_t CpuemShim::getCuAddRange(uint32_t cu_index) {
+    uint64_t cuAddRange = 64 * 1024;
+    for (const auto& cuInfo : mCURangeMap) {
+      std::string instName = cuInfo.first;
+      int cuIdx = static_cast<int>(cu_index);
+      int tmpCuIdx = xclIPName2Index(instName.c_str());
+      mLogStream << __func__ << " , instName :  " << instName  << " cuIdx : " << cuIdx << " tmpCuIdx: " << tmpCuIdx << std::endl;
+      if (tmpCuIdx == cuIdx) {
+        cuAddRange = cuInfo.second;
+        mLogStream << __func__ << " , cuAddRange :  " << cuAddRange << std::endl;
+      }  
+    }
+    return cuAddRange;	  
+  }
+
+  bool CpuemShim::isValidOffset(uint32_t offset, uint64_t cuAddRange) { 
+    if (offset >= cuAddRange || (offset & (sizeof(uint32_t) - 1)) != 0) {
+      std::string strMsg = "ERROR: [SW-EMU 21] xclRegRW - invalid CU offset: " + std::to_string(offset);
+      mLogStream << __func__ << strMsg << std::endl;
+      return false;
+    }
+    return true;
+  }
+
+  int CpuemShim::xclRegRW(bool rd, uint32_t cu_index, uint32_t offset, uint32_t *datap)
+  {
+    if (mLogStream.is_open())
+      mLogStream << __func__ << ", " << std::this_thread::get_id() << ", " << "CU Idx : " << cu_index << " Offset : " << offset << " Datap : " << (*datap) << std::endl;
+
+    if (!isValidCu(cu_index))
+      return -EINVAL;
+
+    const auto& cuidx2addr = mCoreDevice->get_cus();
+
+    uint64_t cuAddRange = getCuAddRange(cu_index);
+
+    if (!isValidOffset(offset, cuAddRange))
+      return -EINVAL;	
+
+    const unsigned REG_BUFF_SIZE = 0x4;
+    std::array<char, REG_BUFF_SIZE> buff;
+    uint64_t baseAddr = cuidx2addr[cu_index];
+    if (rd) {
+      size_t size=4;
+      xclRegRead_RPC_CALL(xclRegRead,baseAddr,offset,buff.data(),size,0,0);
+      auto tmp_buff = reinterpret_cast<uint32_t*>(buff.data());
+      *datap = tmp_buff[0];
+    }
+    else {
+      uint32_t * tmp_buff = reinterpret_cast<uint32_t*>(buff.data());
+      tmp_buff[0] = *datap;
+      xclRegWrite_RPC_CALL(xclRegWrite,baseAddr,offset,tmp_buff,0,0);
+    }	
+    return 0;
```

Comment:
Hmm, why not just use a uint32_t for buff and pass the address of buff to the RPC calls? E.g. something like this:
```
uint32_t buff;   
...
if (rd) {
  ...
  xclRegRead_RPC_CALL(,,,&buff, sizeof(buff), 0, 0);
}
...
```
You could even just use the `datap` directory assuming it is not null, which isn't checked anyway (in the read case).

Reply:
Hi,
Since the xclRegRead_RPC_CALL accepts  bytes, I have used std::array<char, REG_BUFF_SIZE> buff;
I am seeing compilation issues, when we are using as uint32_t.
I have accidentally closed this PR. I have recreated a new one with similar changes.